base/compiler/abstractinterpretation.jl

```diff
@@ -3002,6 +3003,8 @@ function typeinf_local(interp::AbstractInterpreter, frame::InferenceState)
                         condt = Conditional(condx, Const(true), Const(false))
                     end
                     condval = maybe_extract_const_bool(condt)
+                    nothrow = (condval !== nothing) || ⊑(𝕃ᵢ, orig_condt, Bool)
+                    nothrow && add_curr_ssaflag!(frame, IR_FLAG_NOTHROW)
```

Comment:
There's some additional issues here where the `merge_effects!` down below isn't in the right place, but I'm already fixing that in #50805.